module/os/linux/zfs/zvol_os.c

```diff
@@ -195,27 +330,32 @@ static void
 zvol_discard(zv_request_t *zvr)
 {
 	struct bio *bio = zvr->bio;
+	struct request *rq = zvr->rq;
 	zvol_state_t *zv = zvr->zv;
-	uint64_t start = BIO_BI_SECTOR(bio) << 9;
-	uint64_t size = BIO_BI_SIZE(bio);
+	uint64_t start = io_offset(bio, rq);
+	uint64_t size = io_size(bio, rq);
 	uint64_t end = start + size;
 	boolean_t sync;
 	int error = 0;
 	dmu_tx_t *tx;
+	struct request_queue *q = zv->zv_zso->zvo_queue;
+	struct gendisk *disk = zv->zv_zso->zvo_disk;
+	unsigned long start_time = 0;
+
+	boolean_t acct = blk_queue_io_stat(q);
 
 	ASSERT3P(zv, !=, NULL);
```

Comment:
This ASSERT is no longer useful since we'll have already dereferenced `zv` above.  I'd just move the variable declarations back to after the ASSERT.  The same is true for `zvol_write()`, `zvol_read()`, etc.